File: 백준/Silver/5212. 지구 온난화/지구 온난화.cc
#include <iostream>

using namespace std;

char arr[13][13];
char forPrint[13][13];

int dx[4] = {-1, 0, 1, 0};
int dy[4] = {0, 1, 0, -1};

int N, M;

inline void Search() {

	int ix = 11;
	int ax = -1;
	int iy = 11;
	int ay = -1;

	for (int i = 0; i < N; i++) {
		for (int j = 0; j < M; j++) {
			char target = arr[i][j];

			if (target == 'X') {
				int cnt = 0;

				for (int d = 0; d < 4; d++) {
					int nx = i + dx[d];
					int ny = j + dy[d];

					if (0 > nx || 0 > ny || N <= nx || M <= ny || arr[nx][ny] == '.') {
						continue;
					}
					cnt++;
				}

				if (cnt >= 2) {
					forPrint[i][j] = 'X';
					ax = max(ax, i);
					ix = min(ix, i);
					ay = max(ay, j);
					iy = min(iy, j);
				}
				else {
					forPrint[i][j] = '.';
				}
			}
			else {
				forPrint[i][j] = '.';
			}
		}
	}

	for (int i = ix; i <= ax; i++) {
		for (int j = iy; j <= ay; j++) {
			cout << forPrint[i][j];
		}
		cout << '\n';
	}

	return;
}



int main() {

	cin >> N >> M;

	for (int i = 0; i < N; i++) {
		for (int j = 0; j < M; j++) {
			cin >> arr[i][j];
		}
	}

	Search();

	return 0;
}

File: 백준/Silver/5212. 지구 온난화/지구 온난화.py
from collections import deque

dx = [-1, 0, 1, 0]
dy = [0, 1, 0, -1]

N, M = map(int, input().split())
arr = [list(input()) for _ in range(N)]

global_visit = [[False]*M for i in range(N)]

q = deque()

for i in range(N):
    for j in range(M):
        if arr[i][j] == 'X':
            q.append((i, j))
            global_visit[i][j] = True

step = 0
curSize = 0

while q:
    curSize = q.__len__()

    for c in range(curSize):
        target = q.popleft()
        cnt = 0
        for i in range(4):
            nx = target[0] + dx[i]
            ny = target[1] + dy[i]
            if nx < 0 or ny < 0 or nx >= N or ny >= M: continue
            if arr[nx][ny] == 'X':
                cnt += 1

        if cnt < 2:
            global_visit[target[0]][target[1]] = False


max_x = -1
min_x = 100
max_y = -1
min_y = 100

for i in range(N):
    for j in range(M):
        if global_visit[i][j] == True:
            max_x = max(max_x, i)
            min_x = min(min_x, i)

            max_y = max(max_y, j)
            min_y = min(min_y, j)
        elif global_visit[i][j] == False:
            arr[i][j] = '.'

for i in range(min_x, max_x+1):
    for j in range(min_y, max_y+1):
        print(arr[i][j], end='')
    print()